### ISSUE TITLE
Event-driven task triggering

### ISSUE THREAD

**Issue by hjoliver** [cylc/cylc-flow#3741]
Post SoD we could easily do event-driven triggering (like we already do for event-driven finished-task removal, for instance): when a parent-task output is completed and the dependent child spawned, immediately check if the child's prerequisites are satisfied and if so immediately queue for submission.  This would avoid another task pool iteration loop (find all tasks that are ready, and queue them then).

However I held off doing this during SoD implementation because I thought it would probably break our job submission batching which is important for large suites with task families whose members all become ready to run at once.

Internal queues probably provide the answer: event-driven queuing of satisfied tasks, then batch-submission of currently queued tasks.  However, internal queues are not currently implemented as actual queues - they're just dicts that record information about tasks that actually remain in the main task pool.  Re-implement queues as queues?

**Comment by oliver-sanders** [comment]
Using a queue to decouple job submission from the main-loop would be highly desirable.

The job submission component should be an async loop which runs in parallel with the main loop i.e:

```python
# scheduler.py

asyncio.gather(
    self.main_loop,
    self.submitter.loop,
)
```

To keep job batching we can do something like:

```python
# submitter.py

async def loop()
    while True:
        batch = await self.get_batch()
        await asyncio.gather(map(self.submit, batch))
        await asyncio.sleep(self.sleep_time)  # wait for a period of time to allow submissions to pile up
```

This should be fairly straight forward, the main issue will likely be the functional tests which often rely on assumptions about the order of events in the main loop. Assumptions which this will completely shatter by allowing job submission to happen asynchronously with the main loop.
